pkg/pool-manager/virtualmachine_pool.go

```diff
@@ -214,23 +207,22 @@ func (p *PoolManager) UpdateMacAddressesForVirtualMachine(previousVirtualMachine
 		}
 	}
 
-	// Release delta interfaces
-
-	logger.V(1).Info("delta interfaces to release",
-		"interfaces Map", deltaInterfacesMap)
-	p.releaseMacAddressesFromInterfaceMap(deltaInterfacesMap)
-
-	// Release old allocations
-	logger.V(1).Info("old interfaces to release",
-		"interfaces Map", releaseOldAllocations)
-	p.releaseMacAddressesFromInterfaceMap(releaseOldAllocations)
-
 	// Update only new macs in VmToMacPoolMap. we must do this in the webhook context to avoid allocation races.
 	err := p.updateVmToMacPoolMap(macChanges{allocations: newAllocations}, vmNamespaced(copyVM))
 	if err != nil {
 		return err
 	}
 
+	// released macs are marked here for deletion in macmap, and will be deleted in the controller should the webhook chain is persisted.
+	// Until then, the to-be released macs are kept in the VmToMacPoolMap.
+	newReleases := mergeMaps(deltaInterfacesMap, releaseOldAllocations)
+	if len(newReleases) != 0 {
+		// Mark macs for deletion, will be released in the controller reconcile
+		logger.V(1).Info("macs marked for release",
+			"delta interfaces", deltaInterfacesMap, "old interfaces", releaseOldAllocations)
```

Comment:
As @qinqon mentions below, the `deltaInterfacesMap` probably contains also the content of `releaseOldAllocations`

Reply:
yes I see what you mean, it's probably safer to deep copy